src/mds/Locker.cc

```diff
@@ -900,16 +900,26 @@ void Locker::drop_non_rdlocks(MutationImpl *mut, set<CInode*> *pneed_issue)
     issue_caps_set(*pneed_issue);
 }
 
-void Locker::drop_rdlocks_for_early_reply(MutationImpl *mut)
+void Locker::handle_locks_for_early_reply(MutationImpl *mut)
 {
   set<CInode*> need_issue;
+  bool nudged = false;
+
+  dout(10) << __func__ << ": " << *mut << dendl;
 
   for (auto it = mut->locks.begin(); it != mut->locks.end(); ) {
+    SimpleLock *lock = it->lock;
+    if (!nudged) {
+      /* A request may finally early reply only after another request has
+       * triggered an unstable state change. We need to nudge the log now to
+       * move things along.
+       */
+      nudged = nudge_log(lock);
```

Comment:
Copying here our discussion from private chat

> Does it suffice to nudge once and only for the first nudgable lock? ISTM, the mdlog flush is that underlying requirement, yes?

Yes. There's no reason to send multiple flushes; it may only create more work for the flush thread.

> Morevoer, in the past, such things have been addressed by nudging the lock during the actual state transition for a lock. E.g.: https://github.com/ceph/ceph/pull/45243

getattr does nudge the lock but the code wrongly assumes that whatever request holds the lock has already early replied.